pkg/controller/controlplane/valuesprovider.go

```diff
@@ -457,17 +458,90 @@ func (vp *valuesProvider) GetControlPlaneShootCRDsChartValues(
 func (vp *valuesProvider) GetStorageClassesChartValues(
 	_ context.Context,
-	_ *extensionsv1alpha1.ControlPlane,
+	controlPlane *extensionsv1alpha1.ControlPlane,
 	cluster *extensionscontroller.Cluster,
 ) (map[string]interface{}, error) {
 	k8sVersionLessThan119, err := version.CompareVersions(cluster.Shoot.Spec.Kubernetes.Version, "<", openstack.CSIMigrationKubernetesVersion)
 	if err != nil {
 		return nil, err
 	}
+	k8sVersionLessThan112, err := version.CompareVersions(cluster.Shoot.Spec.Kubernetes.Version, "<", "1.12")
+	if err != nil {
+		return nil, err
+	}
 
-	return map[string]interface{}{
-		"useLegacyProvisioner": k8sVersionLessThan119,
-	}, nil
+	providerConfig := api.CloudProfileConfig{}
+	if cluster.CloudProfile.Spec.ProviderConfig != nil {
+		if _, _, err := vp.Decoder().Decode(cluster.CloudProfile.Spec.ProviderConfig.Raw, nil, &providerConfig); err != nil {
+			return nil, errors.Wrapf(err, "could not decode providerConfig of controlplane '%s'", kutil.ObjectName(controlPlane))
+		}
+	}
+	values := make(map[string]interface{})
+	if providerConfig.StorageClasses != nil && len(providerConfig.StorageClasses) != 0 {
+		allSc := make([]map[string]interface{}, len(providerConfig.StorageClasses))
+		for i, sc := range providerConfig.StorageClasses {
+			var storageClassValues = map[string]interface{}{
+				"name": sc.Name,
+			}
+
+			if sc.Default != nil && *sc.Default {
+				storageClassValues["default"] = true
+			}
+
+			if len(sc.Annotations) != 0 {
+				storageClassValues["annotations"] = sc.Annotations
+			}
+			if len(sc.Labels) != 0 {
+				storageClassValues["labels"] = sc.Labels
+			}
+			if len(sc.Parameters) != 0 {
+				storageClassValues["parameters"] = sc.Parameters
+			}
+
+			storageClassValues["provisioner"] = openstack.CSIStorageProvisioner
+			if sc.Provisioner != nil && *sc.Provisioner != "" {
+				storageClassValues["provisioner"] = sc.Provisioner
+			}
+
+			if sc.ReclaimPolicy != nil && *sc.ReclaimPolicy != "" {
+				storageClassValues["reclaimPolicy"] = sc.ReclaimPolicy
+			}
+
+			if sc.VolumeBindingMode != nil && *sc.VolumeBindingMode != "" {
+				storageClassValues["volumeBindingMode"] = sc.VolumeBindingMode
+			}
+
+			allSc[i] = storageClassValues
+		}
+		values["storageclasses"] = allSc
+		return values, err
+	}
+
+	storageclasses := []map[string]interface{}{
+		{
+			"name":              "default",
+			"default":           true,
+			"provisioner":       openstack.CSIStorageProvisioner,
+			"volumeBindingMode": storagev1.VolumeBindingImmediate,
+		},
+		{
+			"name":              "default-class",
+			"provisioner":       openstack.CSIStorageProvisioner,
+			"volumeBindingMode": storagev1.VolumeBindingImmediate,
+		}}
+
+	if k8sVersionLessThan119 {
+		if k8sVersionLessThan112 {
+			storageclasses[0]["volumeBindingMode"] = storagev1.VolumeBindingImmediate
+			storageclasses[1]["volumeBindingMode"] = storagev1.VolumeBindingWaitForFirstConsumer
```

Comment:
For all the defaults that we currently have we use always `WaitForFirstConsumer` as binding mode, now we switch to `Immediate`. As this change affects only newly created volumes I guess we should switch for all defaults to `Immediate`.
@kon-angelo wdyt?

Reply:
I would also prefer to have it as `WaitForFirstConsumer` to avoid having the volumes to be created in the wrong zone.

Reply:
I changed all to `WaitForFirstConsumer` and also removed the now unused check for older k8s versions.